turbo/mixins.py

```diff
@@ -1,53 +1,58 @@
-from asgiref.sync import async_to_sync
-
-from channels.layers import get_channel_layer
+from django.db.models import Model
 
+import turbo
 from turbo import (
-    channel_name_for_instance,
     CREATED,
     UPDATED,
     DELETED,
     REPLACE,
     REMOVE,
-    APPEND,
-)
+    APPEND, )
 
 
 class BroadcastableMixin(object):
+    default_stream_action = APPEND
+    turbo_streams_template = None
+
+    def get_turbo_streams_template(self, target):
+        raise NotImplementedError
+
+    def get_dom_target(self, target):
+        raise NotImplementedError
+
+    def append_context(self, target):
+        return {}
+
+    def _get_context(self, target):
+        context = {"model_template": self.get_turbo_streams_template(target)}
+        context.update(self.append_context(target))
+        return context
+
+    def send_broadcast(self, target, action):
+        turbo.send_broadcast(target, self.get_dom_target(target), action,
+                             self.get_turbo_streams_template(target), self._get_context(target))
+
+
+class BroadcastableModelMixin(BroadcastableMixin):
     broadcasts_to = []  # Foreign Key fieldnames to broadcast updates for.
     broadcast_self = True  # Whether or not to broadcast updates on this model's own stream.
     inserts_by = APPEND  # Whether to append or prepend when adding to a list (broadcasting to a foreign key).
-    turbo_streams_template = None
 
-    def get_turbo_streams_template(self):
+    def get_turbo_streams_template(self, target):
         if self.turbo_streams_template is not None:
             return self.turbo_streams_template
         app_name, model_name = self._meta.label.lower().split(".")
         return f"{app_name}/{model_name}.html"
 
-    def broadcast_to_instance(self, instance, action, to_list, partial=None, context=None):
-        if context is None:
-            context = dict()
-        if partial is None:
-            partial = self.get_turbo_streams_template()
-
-        channel_layer = get_channel_layer()
-        channel_name = channel_name_for_instance(instance)
-        async_to_sync(channel_layer.group_send)(
-            channel_name,
-            {
-                "type": "notify",
-                "model": self._meta.label,
-                "pk": self.pk,
-                "action": action,
-                "channel_name": channel_name,
-                "template": partial,
-                "context": context,
-                "to_list": to_list,
-            },
-        )
+    def append_context(self, target):
+        # Add Custom context
+        app_name, model_name = self._meta.label.lower().split(".")
+        return {
+            "object": self,
```

Comment:
This may work with the in-memory channel layer that channels uses in development, but I'm pretty sure this will produce some wacky results using Redis in production. It might even error -- pretty sure the channel layer will call `json.dumps`.